### ISSUE TITLE
[BUG] Spans from flag evaluation not being created in Java

### ISSUE THREAD

**Issue by julianocosta89** [open-feature/flagd#1369]
### Observed behavior

When updating the dependencies for the AdService in the OTel demo, I've noticed that the flagd spans are not being created in Java.
We do have the flagd spans connected with RecommendationService (Python), CartService (.NET) and Frontend-proxy (Envoy).

But for AdService no.
Even though, we follow the recommended approach documented here: https://flagd.dev/providers/java/#opentelemetry-tracing-rpc-only

https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/adservice/src/main/java/oteldemo/AdService.java#L86-L94

### Expected Behavior

Whenever AdService checks the state of the FeatureFlag, I'd expect 3 spans from flagd to be created, as it is happening with other services.

### Steps to reproduce

```
git clone git@github.com:open-telemetry/opentelemetry-demo.git
cd opentelemetry-demo
docker compose up -d
```

Navigate to Jaeger UI: <http://localhost:8080/jaeger/ui/>

Check the received traces from AdService and Flagd.

**Comment by Kavindu-Dodan** [comment]
@julianocosta89 I had a look into this. The service use `withGlobalTelemetry(true)` which internally obtains an `OpenTelemetry` instance from `GlobalOpenTelemetry.get()`. This follows Java manual instrumentation guidelines [1]. And in the AdService, there is no manual configuration initializations and this is the root cause for this observation. 

[1] - https://opentelemetry.io/docs/languages/java/instrumentation/#manual-configuration

**Comment by julianocosta89** [comment]
@Kavindu-Dodan we are using the OTel Java agent, isn't that done automatically?
We have another service, written in Kotlin, and for that it is working fine.
https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/frauddetectionservice/src/main/kotlin/frauddetectionservice/main.kt#L32-L36

**Comment by julianocosta89** [comment]
I've done some tests again and it seems that it is actually creating spans, but for some reason just once.
Here is the trace containing the flagd spans:
![Screenshot 2024-07-25 at 12 21 23](https://github.com/user-attachments/assets/d0fa4e64-b049-4700-9790-88c041016a22)

Here are the metrics from one of the flags from AdService:
![Screenshot 2024-07-25 at 12 20 43](https://github.com/user-attachments/assets/d9638fbf-f97a-4bbb-9fb5-4df79a0fe817)


As we can see, the flag is being queried every time AdService is called, but the spans from flagd are just arriving once.
Maybe we have some error in the implementation of AdService.

**Comment by beeme1mr** [comment]
Hey @julianocosta89, this is an optimization in the Java provider that can disabled. Basically, a value can be cached if it's "static" and the provider is able to establish a persistent connection with flagd to bust the cache when the flag configuration changes.

Here are the configuration options that are available.
https://github.com/open-feature/java-sdk-contrib/tree/main/providers/flagd#configuration-options

**Comment by julianocosta89** [comment]
> Hey @julianocosta89, this is an optimization in the Java provider that can disabled. Basically, a value can be cached if it's "static" and the provider is able to establish a persistent connection with flagd to bust the cache when the flag configuration changes.
> 
> Here are the configuration options that are available. https://github.com/open-feature/java-sdk-contrib/tree/main/providers/flagd#configuration-options

Oh, cool TIL.

Any reason the behaviour is different for Java and Kotlin?
Both are using the same dependencies and none have the cache disabled.

**Comment by beeme1mr** [comment]
There are a few reasons why a value wouldn't be cashed. For example, if there's a targeting rule in the flag configuration. I'll take a look tomorrow.

**Comment by beeme1mr** [comment]
Hey @julianocosta89, can you confirm that the `kafkaQueueProblems` doesn't have a targeting rule defined in your environment? That would immediately make it ineligible to be cached. I didn't notice any obvious reasons this isn't being cached, but I didn't have a chance to investigate in depth.

**Comment by julianocosta89** [comment]
this is what I currently have for `kafkaQueueProblems`:

```
"kafkaQueueProblems": {
  "description": "Overloads Kafka queue while simultaneously introducing a consumer side delay leading to a lag spike",
  "state": "ENABLED",
  "variants": {
    "on": 100,
    "off": 0
  },
  "defaultVariant": "off"
},
```


And this is the kotlin snippet:

```kotlin
fun main() {
    val options = FlagdOptions.builder()
    .withGlobalTelemetry(true)
    .build()
    val flagdProvider = FlagdProvider(options)
    OpenFeatureAPI.getInstance().setProvider(flagdProvider)

    val props = Properties()
    props[KEY_DESERIALIZER_CLASS_CONFIG] = StringDeserializer::class.java.name
    props[VALUE_DESERIALIZER_CLASS_CONFIG] = ByteArrayDeserializer::class.java.name
    props[GROUP_ID_CONFIG] = groupID
    val bootstrapServers = System.getenv("KAFKA_SERVICE_ADDR")
    if (bootstrapServers == null) {
        println("KAFKA_SERVICE_ADDR is not supplied")
        exitProcess(1)
    }
    props[BOOTSTRAP_SERVERS_CONFIG] = bootstrapServers
    val consumer = KafkaConsumer<String, ByteArray>(props).apply {
        subscribe(listOf(topic))
    }

    var totalCount = 0L
    consumer.use {
        while (true) {
            totalCount = consumer
                .poll(ofMillis(100))
                .fold(totalCount) { accumulator, record ->
                    val newCount = accumulator + 1
                    if (getFeatureFlagValue("kafkaQueueProblems") > 0) {
                        logger.info("FeatureFlag 'kafkaQueueProblems' is enabled, sleeping 1 second")
                        Thread.sleep(1000)
                    }
                    val orders = OrderResult.parseFrom(record.value())
                    logger.info("Consumed record with orderId: ${orders.orderId}, and updated total count to: $newCount")
                    newCount
                }
        }
    }
}

/**
* Retrieves the status of a feature flag from the Feature Flag service.
*
* @param ff The name of the feature flag to retrieve.
* @return `true` if the feature flag is enabled, `false` otherwise or in case of errors.
*/
fun getFeatureFlagValue(ff: String): Int {
    val client = OpenFeatureAPI.getInstance().client
    // TODO: Plumb the actual session ID from the frontend via baggage?
    val uuid = UUID.randomUUID()

    val clientAttrs = mutableMapOf<String, Value>()
    clientAttrs["session"] = Value(uuid.toString())
    client.evaluationContext = ImmutableContext(clientAttrs)
    val intValue = client.getIntegerValue(ff, 0)
    return intValue
}
```

**Comment by toddbaert** [comment]
@julianocosta89 could it be that you are evaluating the same flag in quick succession, and evaluations are interleaving? The flag value can't be cached until the RPC completes, so if 2 evaluations happen quickly, they will both cause RPCs since theirs no previously resolved value in the cache (yet)..

Besides some transient connection issue, I can't see another reason. If the `reason` field is coming back as `STATIC`, subsequent evaluations should result in cache hits with your configuraiton :thinking:

**Comment by aepfli** [comment]
I am doing a little bit of spring issue cleaning, @julianocosta89 is this issue still present?

**Comment by julianocosta89** [comment]
Yes, the 3 feature flags we have for `Ad` are being queried constantly, as we can see on the evaluation metric:

![Image](https://github.com/user-attachments/assets/57ccd025-408c-45bf-807b-15fab9c33049)

But on Traces, this happens only once:

![Image](https://github.com/user-attachments/assets/e6701e4c-8184-4857-8bf8-77f419723322)

All other Traces do not have the `flagd` spans.

One thing that I also noticed while double checking was that I'm getting a couple of traces with 10min duration,and 31 span events 🤯

Maybe those 2 cases are interconnected:

![Image](https://github.com/user-attachments/assets/1943529c-cbf5-4d31-81f9-dc4f95e1b154)

**Comment by julianocosta89** [comment]
> Hey [@julianocosta89](https://github.com/julianocosta89), the 10-minute traces are for the event stream to watch for flag change events. This is expected behavior but we may want to adjust how it's monitored in the OTel demo. By the way, will you be at Contribfest? A few OpenFeature maintainers were planning on attending and it may be a good opportunity to work on some improvements.

Sounds great, I'll be there and I'm planning on spending a lot of time on the Observatory as well!
Hope to meet you there 😊
